browser/ui/views/playlist/playlist_bubble.cc

```diff
@@ -0,0 +1,56 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/playlist/playlist_bubble.h"
+
+#include <utility>
+
+#include "brave/browser/ui/views/playlist/playlist_action_icon_view.h"
+#include "brave/components/playlist/browser/playlist_tab_helper.h"
+#include "ui/base/metadata/metadata_impl_macros.h"
+#include "ui/views/bubble/bubble_border.h"
+
+namespace playlist {
+
+PlaylistBubble::PlaylistBubble(views::View* anchor_view,
+                               base::WeakPtr<PlaylistTabHelper> tab_helper)
+    : BubbleDialogDelegateView(anchor_view,
+                               views::BubbleBorder::Arrow::TOP_RIGHT),
+      tab_helper_(std::move(tab_helper)) {
+  CHECK(anchor_view);
+  CHECK(tab_helper_);
+
+  auto* controller = PlaylistBubblesController::FromWebContents(
+      &tab_helper_->GetWebContents());
+  CHECK(controller);
+  controller_ = controller->AsWeakPtr();
+}
+
+PlaylistBubble::~PlaylistBubble() = default;
+
+void PlaylistBubble::OnWidgetDestroyed(views::Widget*) {
+  if (!controller_) {
+    return;
+  }
+
+  controller_->OnBubbleClosed();
+
+  if (auto* anchor_view = GetAnchorView();
+      anchor_view &&
+      next_bubble_ != PlaylistBubblesController::BubbleType::kInfer) {
+    // Post as task, otherwise
+    // "|anchor_view| has already anchored a focusable widget."
+    base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+        FROM_HERE,
+        base::BindOnce(
+            &PlaylistBubblesController::ShowBubble, controller_,
+            static_cast<PlaylistActionIconView*>(anchor_view)->AsWeakPtr(),
```

Comment:
How about replace static_cast<> with View::AsViewClass<>() and check the invariant? Or changing PlaylistBubble's first param from `View` => `PlaylistActionIconView` makes the invariant more explicit.

Reply:
You mean the constructor's first param? That wouldn't help, since `BubbleDialogDelegate::GetAnchorView()` would still give you a `views::View*`. Going through the `ClassMetaData` stuff wouldn't buy us too much here, as we know that `Playlist` has only one action icon (and most likely won't have another one ever), consequently, the `static_cast<>` is a safe down-cast — no branching/`CHECK()`ing required (`views::AsViewClass<PlaylistActionIconView>(anchor_view)` would eventually perform the same exact cast by the way).

Reply:
What I was thinking is `Easy to use, hard to misuse`. At the moment, it's so apparent that the AnchorView would be the PlaylistActionIconView. But PlaylistBubble still can be constructed with other anchor view, which makes static_cast<> to PlaylistActionIconView makes not rock solid. So I thought we can make the invariant more explicit by
* convert the first parma of constructor and leave static_cast<> as is.
* Use View::AsViewClass<> and check the invariant.

Reply:
If we want to go the extra pedantic route, then let's go with:
```cpp
auto* action_icon_view =
    views::AsViewClass<PlaylistActionIconView>(anchor_view);
CHECK(action_icon_view);
```
since the anchor view can be reset.